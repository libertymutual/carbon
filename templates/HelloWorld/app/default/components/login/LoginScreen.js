import React from 'react';
import PropTypes from 'prop-types';
import { Button, ScrollView, StyleSheet, Text, TextInput, View } from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import AuthActions from '../../actions/AuthActions';
import configs from './authConfigs.json';
const { Auth } = require('NativeModules');

const fieldNames = {
  username: "username",
  password: "password",
  authUrl: "auth url",
  clientId: "client Id",
  clientSecret: "client secret",
  validatorId: "validator Id"
};

/**
* Login Screen
* @extends {React.Component}
*/
export class LoginScreen extends React.Component {
  constructor() {
    super()
    this.state = {
      textOutput: "Auth output:",
    }
   this._updateTextOutput()
  }
  render() {
    return (
      <ScrollView>
        <View style={styles.input}>
          <Text style={styles.fieldLabel}>
            {fieldNames.username}
          </Text>
          <TextInput
            style={styles.textInput}
            placeholder={fieldNames.username}
            value={this.props.auth.username}
            onChangeText={(text) => this.props.authActions.updateUsername(text)}
          />
          <Text style={styles.fieldLabel}>
            {fieldNames.password}
          </Text>
          <TextInput
            style={styles.textInput}
            placeholder={fieldNames.password}
            value={this.props.auth.password}
            onChangeText={(text) => this.props.authActions.updatePassword(text)}
          />
          <Text style={styles.fieldLabel}>
            {fieldNames.authUrl}
          </Text>
          <TextInput
            style={styles.textInput}
            placeholder={fieldNames.authUrl}
            value={this.props.auth.authUrl}
          />
          <Text style={styles.fieldLabel}>
            {fieldNames.clientId}
          </Text>
          <TextInput
            style={styles.textInput}
            placeholder={fieldNames.clientId}
            value={this.props.auth.clientId}
          />
          <Text style={styles.fieldLabel}>
            {fieldNames.clientSecret}
          </Text>
          <TextInput
            style={styles.textInput}
            placeholder={fieldNames.clientSecret}
            value={this.props.auth.clientSecret}
          />
          <Text style={styles.fieldLabel}>
            {fieldNames.validatorId}
          </Text>
          <TextInput
            style={styles.textInput}
            placeholder={fieldNames.validatorId}
            value={this.props.auth.validatorId}
          />

        </View>
        <Button
          onPress={this._useDefaultCredentials}
          title={"Load React Native credentials"}
        />
        <Button
          onPress={this._configureAuth}
          title={"Modify native credentials"}
        />
        <Button
          onPress={this._testOauth2}
          title={"Login (test oauth2)"}
        />
        <Button
          onPress={this._configureFromNativeApp}
          title={"Configure from native authConfigs"}
        />
        <Button
          onPress={this._logout}
          title={"Logout"}
        />
        <View style={styles.textOutput}>
          <Text>
            {this.state.textOutput}
          </Text>
        </View>
      </ScrollView>);
  }

  /**
   * Pull the default credentials from the authConfigs.json file.
   */
  _useDefaultCredentials = () => {
    let defaultConfig = configs[0]

    this.props.authActions.updateAuthUrl(defaultConfig.authUrl)
    this.props.authActions.updateConfigName(defaultConfig.name)
    this.props.authActions.updatePassword(defaultConfig.password)
    this.props.authActions.updateClientId(defaultConfig.clientId)
    this.props.authActions.updateClientSecret(defaultConfig.clientSecret)
    this.props.authActions.updateUsername(defaultConfig.username)
    this.props.authActions.updateValidatorId(defaultConfig.validatorId)
  }

  /**
   * Configure the auth module using the entered credentials
   * This overrides the natively loaded credentials
   */
  _configureAuth = () => {
    Auth.configure(
      this.props.auth.configName,
      this.props.auth.authUrl,
      this.props.auth.clientId,
      this.props.auth.clientSecret,
      this.props.auth.validatorId
    );
  }

  /**
   * Login using the username & password entered & retrieve access token.
   */
  _testOauth2 = () => {
    Auth.login(
      this.props.auth.username,
      this.props.auth.password,
    ).then(() => {
      this._updateTextOutput();
    });
  }

  /**
  * Logout the user, this should delete all Token info in the auth module
  */
  _logout = () => {
    Auth.logout().then(() => {
      this._updateTextOutput();
    });
  }

  /**
   * Create an auth configuration from the native app's 'authConfigs.json' file. This file should contain
   * a config with the name that matches the key that is passed in.
   */
  _configureFromNativeApp = () => {
    Auth.configureWithEnvironmentKey("default").then(() => {
      this._updateTextOutput("Auth configured from native app")
    }).catch((error) => {
      let message = "Error configuring from auth file\n" + error.message
      this.setState({ textOutput: message })
    });
  }
  /**
   * Checks the output of the available calls to the Auth module.
   * If the user is logged in, token information & user name & email should be displayed
   * If the user is logged out, then no information should be found.
   */
  _updateTextOutput = (headingText: String = "Auth Module Output:") => {
    Auth.isAuthenticated().then((isAuthenticated) => {
      var output = headingText + "\nIs Authenticated:" + isAuthenticated
      this.setState({ textOutput: output })
      Auth.getAccessToken().then((tokenString) => {
        if (tokenString !== null) {
          output = output + "\nToken Obtained"
        } else {
          output = output + "\nNo Token"
        }
        Auth.getEmail().then((email) => {
          output = output + "\nEmail: " + email
          Auth.getUsername().then((username) => {
            output = output + "\nUsername: " + username
            Auth.getExpiresIn().then((expiresIn) => {
              output = output + "\nExpiresIn: " + expiresIn
              this.setState({ textOutput: output })
            });
          });
        });
      }).catch(error => {
        output = output + "\n" + error.message
        this.setState({ textOutput: output })
      });
    });
  }
}

/**
* Navigation Options
*/
LoginScreen.navigationOptions = {
  title: "Login",
};

/**
* Prop Types
*/
LoginScreen.propTypes = {
  navigation: PropTypes.object.isRequired,
};

/**
* Maps State To Props
* @returns {Object) Object representing the parts of state
* that the component should subscribe to
*/
function mapStateToProps(state) {
  return {
    auth: state.auth,
  };
}

/**
* Maps Dispatch Actions To Props
* @returns {Object) Object representing the actions
 * that should be exposed to the component via props
*/
function mapDispatchToProps(dispatch) {
  return {
    authActions: bindActionCreators(AuthActions, dispatch),
  };
}

/**
* Default export wrapped in a connect to expose state & actions via props
*/
export default connect(mapStateToProps, mapDispatchToProps)(LoginScreen);

/**
* Component Styles
*/
const whiteColor = 'white'
const grayColor = 'gray'
const styles = StyleSheet.create({
  input: {
    margin: 20,
  },
  fieldLabel: {
    fontSize: 14,
    marginLeft: 2
  },
  textInput: {
    flex: 3,
    height: 40,
    backgroundColor: whiteColor,
    borderColor: grayColor,
    borderWidth: 1,
    margin: 2,
    marginBottom: 6,
    paddingLeft: 10,
  },
  textOutput: {
    margin: 4,
    padding: 4,
    borderWidth: 1,
  }
});
