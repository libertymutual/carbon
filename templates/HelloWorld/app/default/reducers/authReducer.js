import * as ActionTypes from '../actions/ActionTypes';

/**
 * Initial Reducer State
 */
export const initialState = {
  isLoggedIn: false,
  username: "",
  password: "",
  authUrl: "",
  clientId: "",
  clientSecret: "",
  configName: "",
  validatorId: "",
}

/**
 * Reducer handling auth related state
 * @param {Object} state The current reducer state
 * @param {Object} action The action event that has been triggered
 * @returns {Object} The updated state
 */
export default function authReducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.LOGIN:
      return { ...state, isLoggedIn: true };
    case ActionTypes.LOGOUT:
      return { ...state, isLoggedIn: false };
    case ActionTypes.AUTH_UPDATE_AUTH_URL:
      return { ...state, authUrl: action.authUrl };
    case ActionTypes.AUTH_UPDATE_CLIENT_ID:
      return { ...state, clientId: action.clientId };
    case ActionTypes.AUTH_UPDATE_CLIENT_SECRET:
      return { ...state, clientSecret: action.clientSecret };
    case ActionTypes.AUTH_UPDATE_CONFIG_NAME:
      return { ...state, configName: action.configName };
    case ActionTypes.AUTH_UPDATE_PASSWORD:
      return { ...state, password: action.password };
    case ActionTypes.AUTH_UPDATE_USERNAME:
      return { ...state, username: action.username };
    case ActionTypes.AUTH_UPDATE_VALIDATOR_ID:
      return { ...state, validatorId: action.validatorId };
    default:
      return state;
  }
}
