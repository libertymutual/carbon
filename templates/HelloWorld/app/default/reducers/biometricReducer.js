import * as ActionTypes from '../actions/ActionTypes';

/**
 * Initial Reducer State
 */
export const initialState = {
    username: "TestUsername",
    password: "TestPassword",
    credentialsSaved: false
};

/**
 * Reducer handling auth related state
 * @param {Object} state The current reducer state
 * @param {Object} action The action event that has been triggered
 * @returns {Object} The updated state
 */
export default function biometricReducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.UPDATE_USERNAME:
      return { ...state, username: action.username };
    case ActionTypes.UPDATE_PASSWORD:
      return { ...state, password: action.password };  
    case ActionTypes.SAVE_CREDENTIALS:
      return { ...state, credentialsSaved: true };
    case ActionTypes.DELETE_CREDENTIALS:
      return { ...state, credentialsSaved: false };
    default:
      return state;
  }
}
