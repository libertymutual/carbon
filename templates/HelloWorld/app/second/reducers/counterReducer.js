import * as ActionTypes from '../actions/ActionTypes';

/**
 * Initial Reducer State
 */
export const initialState = 0;

/**
 * Reducer handling counter related state
 * @param {Object} state The current reducer state
 * @param {Object} action The action event that has been triggered
 * @returns {Object} The updated state
 */
export default function countReducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.INCREMENT:
      return state + 1;
    case ActionTypes.DECREMENT:
      return state - 1;
    default:
      return state;

  }
}
