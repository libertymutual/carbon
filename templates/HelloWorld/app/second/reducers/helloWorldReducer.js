import * as ActionTypes from '../actions/ActionTypes';
import * as Constants from '../constants/Constants';

/**
 * Initial Reducer State
 */
export const initialState = {
  state: Constants.HELLO,
};

/**
 * Reducer handling hello world related state
 * @param {Object} state The current reducer state
 * @param {Object} action The action event that has been triggered
 * @returns {Object} The updated state
 */
export default function helloWorldReducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.HELLO:
      return { ...state, state: Constants.HELLO };
    case ActionTypes.GOODBYE:
      return { ...state, state: Constants.GOODBYE };
    default:
      return state;
  }
}
