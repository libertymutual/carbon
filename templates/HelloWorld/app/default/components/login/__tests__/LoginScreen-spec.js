import 'react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import LoginScreen from '../LoginScreen';
import { initialState } from '../../../reducers/authReducer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

let store;

describe('Login Screen tests', () => {
  beforeEach(() => {
    // Sets up a new store
    store = mockStore({
      auth: initialState
    });
  });

  test('Login Screen renders correctly', () => {
    const renderer = new ShallowRenderer();
    const tree = renderer.render(
      <LoginScreen store={store} navigation={{}} />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
