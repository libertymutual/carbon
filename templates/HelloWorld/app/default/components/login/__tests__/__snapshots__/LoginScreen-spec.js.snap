// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen tests Login Screen renders correctly 1`] = `
<LoginScreen
  auth={
    Object {
      "authUrl": "",
      "clientId": "",
      "clientSecret": "",
      "configName": "",
      "isLoggedIn": false,
      "password": "",
      "username": "",
      "validatorId": "",
    }
  }
  authActions={
    Object {
      "login": [Function],
      "logout": [Function],
      "updateAuthUrl": [Function],
      "updateClientId": [Function],
      "updateClientSecret": [Function],
      "updateConfigName": [Function],
      "updatePassword": [Function],
      "updateUsername": [Function],
      "updateValidatorId": [Function],
    }
  }
  navigation={Object {}}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;
